src/Microsoft.OData.Edm/Csdl/Parsing/AnnotationJsonParser.cs

```diff
@@ -0,0 +1,515 @@
+            {
+                return false;
+            }
+
+            string label = propertyValue.ProcessProperty("$LabeledElementReference", context, (e, c) => e.ParseAsString(c));
+
+            labeledReferenceExp = new CsdlLabeledExpressionReferenceExpression(label, context.Location());
+            return true;
+        }
+
+        /// <summary>
+        /// Parse the record expression expression using <see cref="JsonElement"/>.
+        /// </summary>
+        /// <param name="element">The input JSON element.</param>
+        /// <param name="context">The parser context.</param>
+        /// <returns>the built record expression.</returns>
+        private static CsdlRecordExpression ParseRecordExpression(JsonElement element, JsonParserContext context)
+        {
+            Debug.Assert(context != null);
+
+            // A record expression MAY specify the structured type of its result, which MUST be an entity type or complex type in scope.
+            // If not explicitly specified, the type is derived from the expression�s context
+
+            // The type of a record expression is represented as the @type control information
+            // for example: "@type": "https://example.org/vocabs/person#org.example.person.Manager",
+            // So far, ODL doesn't support the type "@type" with relative path, only supports like "#Model.VipCustomer", or without #
+            // for 4.0, this name MUST be prefixed with the hash symbol (#);
+            // or non-OData 4.0 payloads, built-in primitive type values SHOULD be represented without the hash symbol,
+            // but consumers of 4.01 or greater payloads MUST support values with or without the hash symbol.
+
+            CsdlTypeReference typeReference = null;
+            if (element.TryGetProperty("@type", out JsonElement typeValue))
+            {
+                // Try to build the type. The type should be "Complex type" or "Entity Type".
+                string typeName = typeValue.ProcessProperty("@type", context, (e, c) => e.ParseAsString(c));
+                int index = typeName.IndexOf('#');
+                if (index >= 0)
+                {
+                    typeName = typeName.Substring(index + 1); // remove the "#"
+                }
+
+                typeReference = new CsdlNamedTypeReference(typeName, true, context.Location());
+            }
+
+            IList<CsdlPropertyValue> propertyValues = new List<CsdlPropertyValue>();
+            element.ParseAsObject(context, (propertyName, propertyValue) =>
+            {
+                // skips the @type, because it's processed above.
+                if (propertyName == "@type")
+                {
+                    return;
+                }
+
+                // It MAY contain annotations for itself and its members. Annotations for record members are prefixed with the member name.
+                // So far, it's not supported. So report error all the annotations on record, or just skip it?

```

Comment:
>So report error all the annotations on record, or just skip it [](start = 47, length = 62)

I'm okay reporting a non-fatal error that we didn't read the annotation, but reading annotations should not cause a fatal error (i.e., the parse of the model should still succeed)

Reply:
Got it, that's what i did now. 

JsonUnexpectedElement & JsonUnsupportedElement are both for non-fatal errors.

---
In reply to: [443017531](https://github.com/OData/odata.net/pull/1775#discussion_r443017531) [](ancestors = 443017531)